lib/SPIRV/libSPIRV/SPIRVInstruction.h

```diff
@@ -4225,5 +4225,82 @@ _SPIRV_OP(ConvertHandleToSamplerINTEL)
 _SPIRV_OP(ConvertHandleToSampledImageINTEL)
 #undef _SPIRV_OP
 
+class SPIRVUntypedPrefetchKHR : public SPIRVInstruction {
+public:
+  static const Op OC = OpUntypedPrefetchKHR;
+  static const SPIRVWord FixedWordCount = 3;
+
+  SPIRVUntypedPrefetchKHR(SPIRVType *Ty, std::vector<SPIRVWord> &TheArgs,
+                          SPIRVBasicBlock *BB)
+      : SPIRVInstruction(FixedWordCount, OC, BB) {
+    setHasNoId();
+    setHasNoType();
+    PtrTy = TheArgs[0];
+    NumBytes = TheArgs[1];
+    if (TheArgs.size() > 2)
+      RW.push_back(TheArgs[2]);
+    if (TheArgs.size() > 3)
+      Locality.push_back(TheArgs[3]);
+    if (TheArgs.size() > 4)
+      CacheTy.push_back(TheArgs[4]);
+    assert(BB && "Invalid BB");
+    validate();
+  }
+
+  SPIRVUntypedPrefetchKHR() : SPIRVInstruction(OC) {
+    setHasNoId();
+    setHasNoType();
+  }
+
+  void validate() const override {
+    SPIRVInstruction::validate();
+    assert(getValueType(PtrTy)->isTypePointer());
+    assert(getValueType(PtrTy)->getPointerStorageClass() ==
+           StorageClassCrossWorkgroup);
+    assert(getValueType(NumBytes)->isTypeInt());
+    assert(RW.empty() || (RW.size() == 1 && getValueType(RW[0])->isTypeInt()));
+    assert(Locality.empty() ||
+           (Locality.size() == 1 && getValueType(Locality[0])->isTypeInt()));
+    assert(CacheTy.empty() ||
+           (CacheTy.size() == 1 && getValueType(CacheTy[0])->isTypeInt()));
+  }
+
+  void setWordCount(SPIRVWord TheWordCount) override {
+    SPIRVEntry::setWordCount(TheWordCount);
+    if (TheWordCount > 3)
```

Comment:
I might be missing something, but why do we need these checks?

Reply:
This is the way to handle optional arguments that is used through the translator's codebase. For the current test case without optional parameters these checks are not used, but need to verify if they are  in case we have optional operands.